### PR TITLE
issue 989 added docker-compose for the http-server example with obser…

### PR DESCRIPTION
**Pull request by vigneshp91**
Added docker compose file which run mysql,redis and jaeger and the sample application in docker containers.
Fixes [#989](https://github.com/gofr-dev/gofr/issues/989)

### CODE REVIEW

**Review by aryanmehrotra — CHANGES_REQUESTED** [review]
Can you also run grafana with the dashboard from : https://grafana.com/grafana/dashboards/19905-gofr-application-services-monitoring/

Moreover, we can use gofr tracer instead of jaeger as it is a public api as well.

**Comment by vigneshp91** [comment]
> Can you also run grafana with the dashboard from : https://grafana.com/grafana/dashboards/19905-gofr-application-services-monitoring/
> 
> Moreover, we can use gofr tracer instead of jaeger as it is a public api as well.

I can remove Jaeger but
if we are including Grafana then there will be too many setups has to be done like import dashboard, add datasource from prometheus it will make the docker file complex

**Comment by aryanmehrotra** [comment]
@vigneshp91 yes, but we would want to keep in at least one example so that it becomes reference point for new users to setup

**Comment by vigneshp91** [comment]
@aryanmehrotra I have added the required changes. But the  PKG Unit Testing (v1.21) stage is failing. Not sure because of my change

File: examples/http-server/Dockerfile [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
+# Build stage
+FROM golang:1.22 AS build
 
-RUN mkdir /src/
-WORKDIR /src/
+WORKDIR /src
 COPY . .
 RUN go get ./...
-RUN go build -ldflags "-linkmode external -extldflags -static" -a main.go
+RUN go build -ldflags "-linkmode external -extldflags -static" -a -o /app/main main.go
```
Comment (aryanmehrotra) [inline examples/http-server/Dockerfile:7]: Why not build the image in local-system and just copy the build file and the configs - which would reduce the size of the container.

i.e we can use alpine container?

File: examples/http-server/docker-compose.yaml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      - DB_PORT=3306
+      - DB_DIALECT=mysql
+    ports:
+      - "9000:9000"
+      - "2121:2121"
+    depends_on:
+      - redisdb
+      - mysqldb
+      - grafana
+      - prometheus
```
Comment (aryanmehrotra) [inline examples/http-server/docker-compose.yaml:26]: were you able to see the metrics in the prometheus or grafana dashboard?
as it is not appearing for me

**Review by aryanmehrotra — CHANGES_REQUESTED** [review]

File: examples/http-server/Dockerfile (hunk 1 above)
Comment (aryanmehrotra) [reply examples/http-server/Dockerfile:7]: Lets discuss the approach on this first before actually doing it

File: examples/http-server/docker-compose.yaml (hunk 2 above)
Comment (vigneshp91) [reply examples/http-server/docker-compose.yaml:26]: its visible in grafana dashboard
Screenshot 2024-10-09 at 8 40 23 PM
Screenshot 2024-10-09 at 8 40 57 PM
Screenshot 2024-10-09 at 8 40 57 PM

File: examples/http-server/Dockerfile (hunk 1 above)
Comment (vigneshp91) [reply examples/http-server/Dockerfile:7]: @aryanmehrotra IMO Building the app outside the docker is also a good approach which will reduce the docker build time and simplify the docker file.
Building the app inside the docker also have the advantage that we don't need to depend on the host machine for the dependencies. users can simply install docker and they can start the app.
I'm ok with both the approaches

File: examples/http-server/docker-compose.yaml (hunk 2 above)
Comment (aryanmehrotra) [reply examples/http-server/docker-compose.yaml:26]: my bad, I was trying to test in GitHub code space, probably it had some port forwarding issue.
It works fine in local

File: examples/http-server/Dockerfile (hunk 1 above)
Comment (aryanmehrotra) [reply examples/http-server/Dockerfile:7]: seems fine, lets build in the docker itself if needed will change in future

File: examples/http-server/README.md [hunk 3]
```diff
@@ -4,17 +4,13 @@ This GoFr example demonstrates a simple HTTP server which supports Redis and MyS
...
 ```
 
-- Run the docker image of MySQL
-```console
-docker run --name gofr-mysql -e MYSQL_ROOT_PASSWORD=password -e MYSQL_DATABASE=test -p 2001:3306 -d mysql:8.0.30
-```
+To test the example, follow these steps:
+
+1. Open your browser and navigate to `http://localhost:9000/hello`.
+2. To view the trace in Jaeger, open `http://localhost:16686`.
```
Comment (aryanmehrotra) [inline examples/http-server/README.md:15]: now it won't be visible in jaeger we can update the link in message to view on the gofr tracer website

https://tracer.gofr.dev

**Review by aryanmehrotra — CHANGES_REQUESTED** [review]
1. Can we move prometheus and provisioning in one directory may be keep docker-compose.yml as well there.

It would be clean that way

once this is fixed we can merge this PR.

**Comment by vigneshp91** [comment]
@aryanmehrotra Can we merge this?

**Review by aryanmehrotra — APPROVED** [review]

**Review by Umang01-hash — APPROVED** [review]

**Review by vipul-rawat — APPROVED** [review]

**Pull request merged**